mining/src/mempool/model/frontier.rs

```diff
@@ -0,0 +1,391 @@
+    ///           from `~1/50` (2000 sompi for a transaction with 100K storage mass), to `5M` (100 KAS fee for a transaction with
+    ///           2000 mass = 100·100_000_000/2000), resulting in a range of size 250M (`5M/(1/50)`).
+    ///           By using floating point arithmetics we gain the adjustment of the probability space to the accuracy level required for
+    ///           current samples. And if the space is highly biased, the repeated elimination of top items and the prefix weight computation
+    ///           will readjust it.
+    pub fn sample_inplace<R>(&self, rng: &mut R, policy: &Policy) -> SequenceSelectorInput
+    where
+        R: Rng + ?Sized,
+    {
+        debug_assert!(!self.search_tree.is_empty(), "expected to be called only if not empty");
+
+        // Sample 20% more than the hard limit in order to allow the SequenceSelector to
+        // compensate for consensus rejections.
+        // Note: this is a soft limit which is why the loop below might pass it if the
+        //       next sampled transaction happens to cross the bound
+        let desired_mass = (policy.max_block_mass as f64 * MASS_LIMIT_FACTOR) as u64;
+
+        let mut distr = Uniform::new(0f64, self.total_weight());
+        let mut down_iter = self.search_tree.descending_iter();
+        let mut top = down_iter.next().unwrap();
+        let mut cache = HashSet::new();
+        let mut sequence = SequenceSelectorInput::default();
+        let mut total_selected_mass: u64 = 0;
+        let mut _collisions = 0;
+
+        // The sampling process is converging so the cache will eventually hold all entries, which guarantees loop exit
+        'outer: while cache.len() < self.search_tree.len() && total_selected_mass <= desired_mass {
+            let query = distr.sample(rng);
+            let item = {
+                let mut item = self.search_tree.search(query);
+                while !cache.insert(item.tx.id()) {
+                    _collisions += 1;
+                    if top == item {
+                        // Narrow the search to reduce further sampling collisions
+                        match down_iter.next() {
+                            Some(next) => top = next,
+                            None => break 'outer,
+                        }
+                        let remaining_weight = self.search_tree.prefix_weight(top);
+                        distr = Uniform::new(0f64, remaining_weight);
+                    }
+                    let query = distr.sample(rng);
+                    item = self.search_tree.search(query);
+                }
+                item
+            };
+            sequence.push(item.tx.clone(), item.mass);
+            total_selected_mass += item.mass; // Max standard mass + Mempool capacity bound imply this will not overflow
+        }
+        trace!("[mempool frontier sample inplace] collisions: {_collisions}, cache: {}", cache.len());
```

Comment:
Do you have metrics on the average number of collisions when the mempool size is something like 1M?

Reply:
If the feerate distribution is more or less linear (i.e., linearly distributed over some range), you get ~0 collisions with 1M entries. You start getting collisions only below 100K entries (100K: ~1 collisions, 10K: ~16, 5000: ~33, 2000: ~100; below ~2000 we use the other selector).

The interesting part is that following this question I benchmarked the worst-case distribution and noticed an easy optimization which made a dramatic improvement ([commit](https://github.com/kaspanet/rusty-kaspa/pull/513/commits/18a8fc45349f7cb0e919fd314a67a6ff3eb768f6)) 

The change was to eagerly iterate the down iterator until finding a top item which was not sampled yet. After this, also with worst-case scenarios, the results look good:

```
mempool inplace sampling/inplace sampling worst case (bucket size: 300)
                        time:   [206.51 µs 207.48 µs 208.37 µs]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
---------------------- 
  Avg collisions: 1293
```